substrate/pkg/controller/substrate/subnets.go

```diff
@@ -1,181 +1,137 @@
+	errs := make([]error, len(substrate.Spec.Subnets))
+	workqueue.ParallelizeUntil(ctx, len(substrate.Spec.Subnets), len(substrate.Spec.Subnets), func(i int) {
+		subnets[i], errs[i] = s.ensure(ctx, substrate, substrate.Spec.Subnets[i])
+	})
+	if err := multierr.Combine(errs...); err != nil {
+		return reconcile.Result{}, err
+	}
+	for _, subnet := range subnets {
+		if aws.BoolValue(subnet.MapPublicIpOnLaunch) {
+			substrate.Status.PublicSubnetIDs = append(substrate.Status.PublicSubnetIDs, aws.StringValue(subnet.SubnetId))
 		} else {
-			subnetID = *subnet.SubnetId
+			substrate.Status.PrivateSubnetIDs = append(substrate.Status.PrivateSubnetIDs, aws.StringValue(subnet.SubnetId))
 		}
-		substrate.Status.PublicSubnetIDs = append(substrate.Status.PublicSubnetIDs, subnetID)
 	}
-	return nil
+	return reconcile.Result{}, nil
 }
 
-func subnetExists(existingSubnets []*ec2.Subnet, name, az string) (*ec2.Subnet, bool) {
-	for _, existingSubnet := range existingSubnets {
-		if strings.EqualFold(aws.StringValue(existingSubnet.AvailabilityZone), az) &&
-			strings.EqualFold(subnetName(existingSubnet), name) {
-			return existingSubnet, true
-		}
+func (s *subnets) ensure(ctx context.Context, substrate *v1alpha1.Substrate, subnetSpec *v1alpha1.SubnetSpec) (*ec2.Subnet, error) {
+	subnet, err := s.ensureSubnet(ctx, substrate, subnetName(substrate.Name, subnetSpec.Zone, subnetSpec.Public), subnetSpec.Zone, subnetSpec.CIDR)
+	if err != nil {
+		return nil, err
 	}
-	return nil, false
-}
-
-func subnetName(subnet *ec2.Subnet) string {
-	for _, tag := range subnet.Tags {
-		if aws.StringValue(tag.Key) == "Name" {
-			return aws.StringValue(tag.Value)
-		}
+	routeTableID := substrate.Status.PrivateRouteTableID
+	if subnetSpec.Public {
+		routeTableID = substrate.Status.PublicRouteTableID
+	}
+	if _, err := s.ec2Client.AssociateRouteTableWithContext(ctx, &ec2.AssociateRouteTableInput{RouteTableId: routeTableID, SubnetId: subnet.SubnetId}); err != nil {
+		return nil, fmt.Errorf("associating route table with subnet, %w", err)
 	}
-	return ""
+	logging.FromContext(ctx).Infof("Ensured association of route table %s to subnet %s", aws.StringValue(routeTableID), aws.StringValue(subnet.SubnetId))
+	if !subnetSpec.Public {
+		return subnet, nil
+	}
+	if _, err := s.ec2Client.ModifySubnetAttribute(&ec2.ModifySubnetAttributeInput{SubnetId: subnet.SubnetId, MapPublicIpOnLaunch: &ec2.AttributeBooleanValue{Value: aws.Bool(true)}}); err != nil {
+		return nil, fmt.Errorf("modifying subnet attribute, %w", err)
+	}
+	subnet.MapPublicIpOnLaunch = aws.Bool(true)
+	logging.FromContext(ctx).Infof("Ensured subnet %s is public", aws.StringValue(subnet.SubnetId))
+	return subnet, nil
 }
 
-func (s *subnet) Delete(ctx context.Context, substrate *v1alpha1.Substrate) error {
-	subnets, err := s.getSubnets(ctx, substrate.Name)
+func (s *subnets) ensureSubnet(ctx context.Context, substrate *v1alpha1.Substrate, name string, zone string, cidr string) (*ec2.Subnet, error) {
```

Comment:
nit: we can just have args as `ensureSubnet(ctx context.Context, substrate *v1alpha1.Substrate, name, zone, cidr string)`

Reply:
Sure